drop-transfer/src/file/mod.rs

```diff
@@ -202,58 +221,59 @@ impl File {
         result
     }
 
-    pub fn size(&self) -> u64 {
-        match &self.kind {
-            FileKind::FileToSend { meta, .. } => meta.len(),
-            FileKind::FileToRecv { size } => *size,
-        }
-    }
+    fn walk(path: &Path, config: &DropConfig) -> Result<Vec<Self>, Error> {
+        let parent = path
+            .parent()
+            .ok_or_else(|| crate::Error::BadPath("Missing parent directory".into()))?;
 
-    pub fn id(&self) -> &FileId {
-        &self.file_id
-    }
+        let mut files = Vec::new();
+        let mut breadth = 0;
 
-    pub fn subpath(&self) -> &FileSubPath {
-        &self.subpath
-    }
+        for entry in WalkDir::new(path).min_depth(1).into_iter() {
+            let entry = entry?;
+            let meta = entry.metadata()?;
+
+            if !meta.is_file() {
+                continue;
+            }
 
-    pub fn info(&self) -> Option<FileInfo> {
-        match &self.kind {
-            FileKind::FileToSend { mime_type, .. } => Some(FileInfo {
-                mime_type: mime_type
-                    .as_deref()
-                    .cloned()
-                    .unwrap_or_else(|| "unknown".to_string()),
-                extension: AsRef::<Path>::as_ref(self.subpath.name())
-                    .extension()
-                    .map(|e| e.to_string_lossy().to_string())
-                    .unwrap_or_else(|| "unknown".to_string()),
-                size_kb: (self.size() as f64 / 1024.0).ceil() as i32,
-            }),
-            _ => None,
+            if entry.depth() > config.dir_depth_limit {
+                return Err(Error::TransferLimitsExceeded);
+            }
+
+            breadth += 1;
+
+            if breadth > config.transfer_file_limit {
+                return Err(Error::TransferLimitsExceeded);
+            }
+
+            let subpath = entry
+                .path()
+                .strip_prefix(parent)
+                .map_err(|err| crate::Error::BadPath(err.to_string()))?;
+
+            let subpath = FileSubPath::from_path(subpath)?;
+
+            let path = entry.into_path();
+            let file_id = file_id_from_path(&path)?;
+            let file = Self::new_to_send(subpath, path, meta, file_id)?;
+            files.push(file);
         }
+
+        Ok(files)
     }
 
     // Open the file if it wasn't already opened and return the std::fs::File
     // instance
     pub(crate) fn open(&self, offset: u64) -> crate::Result<FileReader> {
-        match &self.kind {
-            FileKind::FileToSend { meta, source, .. } => {
-                let mut reader = reader::open(source)?;
-                reader.seek(io::SeekFrom::Start(offset))?;
-                FileReader::new(reader, meta.0.clone())
-            }
-            _ => Err(Error::BadFile),
-        }
+        let mut reader = reader::open(&self.source)?;
+        reader.seek(io::SeekFrom::Start(offset))?;
+        FileReader::new(reader, self.meta.0.clone())
```

Comment:
BadFile is not gone in this case, what is the final error happening in case of open failure?

Reply:
disk IO errors, just like before when we passed the `BadFile` check

Reply:
Thanks